featurizers/map4_mhfp_helper_functions.py

```diff
@@ -0,0 +1,140 @@
+"""
+Code inspired by the original work of the authors of the MAP4 Fingerprint:
+https://github.com/reymond-group/map4
+"""
+
+import itertools
+from collections import defaultdict
+
+import numpy as np
+from rdkit.Chem import MolToSmiles, PathToSubmol
+from rdkit.Chem.rdchem import Mol
+from rdkit.Chem.rdmolops import FindAtomEnvironmentOfRadiusN, GetDistanceMatrix
+
+
+def _find_neighborhood(Mol: Mol, idx: int, radius: int) -> str:
+    """
+    Function for getting the neighborhood for given atom. By the neighborhood, we mean the structures,
+    that are adjacent to the given atom.
+    """
+    env = FindAtomEnvironmentOfRadiusN(Mol, idx, radius)
+    atom_map = {}
+
+    submol = PathToSubmol(Mol, env, atomMap=atom_map)
+
+    if idx in atom_map:
+        smiles = MolToSmiles(
+            submol,
+            rootedAtAtom=atom_map[idx],
+            canonical=True,
+            isomericSmiles=False,
+        )
+        return smiles
+    else:
+        return ""
+
+
+def _get_atom_envs(Mol: Mol, radius: int):
+    """
+    For each atom get its environment.
+    """
+    atoms_env = defaultdict(lambda: [])
+    for atom in Mol.GetAtoms():
+        idx = atom.GetIdx()
+        for r in range(1, radius + 1):
+            atoms_env[idx].append(_find_neighborhood(Mol, idx, r))
+
+    return atoms_env
+
+
+def _all_pairs(Mol: Mol, atoms_envs: dict, radius: int, is_counted: bool):
+    """
```

Comment:
Good job with this function, it is really complicated to get algorithmic things like this right